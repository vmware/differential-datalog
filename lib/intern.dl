/* Object interning library.
 * Generates a unique ID for an object, which can be converted back to the original value.
 *
 * This library is more lightweight than the standard `internment.dl` library in
 * that interned object handles are only 32-bit wide (vs 64 bit with
 * `internment.dl`).  In addition, this library does not implement reference
 * counting, so interned object manipulation is more efficient.  On the flip side,
 * it never deallocates interned objects, so unless you expect the set of
 * interned objects to remain constant, you should use `internment.dl` instead.
 */

/* Interned object of type `'A`.
 * While this type is defined for any `'A`, interning is only supported for strings.
 * There is simply no way to obtain an interned object of a different type.
 */
#[size=4]
#[shared_ref]
extern type IObj<'A>

/* Interned string
 */
typedef IString = IObj<string>

/* Intern a string
 */
extern function string_intern(s: string): IString

/* Extract the value of an interned string
 */
extern function istring_str(s: IString): string

/* Returns unique integer identifier of an interned string.
 * Identifiers are assigned in the order strings are fed to the interner.
 */
extern function istring_ord(s: IString): u32
