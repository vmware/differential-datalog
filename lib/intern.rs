use lazy_static::lazy_static;
use std::marker;
use std::thread;
use std::vec;
use std::collections;
use std::sync;
use serde;
use differential_datalog::record::Record;
use differential_datalog::record;

type Symbol = u32;

#[derive(Eq, PartialOrd, PartialEq, Ord, Clone, Hash)]
pub struct intern_IObj<T> {
    pub x: Symbol,
    phantom: marker::PhantomData<T>
}

struct Interner<T> {
    vec: vec::Vec<T>,
    map: collections::HashMap<T, Symbol>
}

impl <T: Clone + Eq + Hash> Interner<T> {
    pub fn new() -> Self { Interner { vec: vec::Vec::new(), map: collections::HashMap::new() } }
    pub fn intern(&mut self, x: &T) -> Symbol {
        if !self.map.contains_key(x) {
            let len = self.map.len() as Symbol;
            self.vec.push(x.clone());
            self.map.insert(x.clone(), len);
            len
        }
        else {
            *self.map.get(x).unwrap()
        }
    }
    pub fn get(&self, sym: Symbol) -> &T {
        &self.vec[sym as usize]
    }
}

type StringInterner = Interner<String>;

lazy_static! {
    static ref global_string_interner: sync::Arc<sync::Mutex<StringInterner>> = sync::Arc::new(sync::Mutex::new(StringInterner::new()));
}

thread_local!(static STRING_INTERNER: sync::Arc<sync::Mutex<StringInterner>> = global_string_interner.clone());

impl Default for intern_IString {
    fn default() -> Self {
        intern_string_intern(&String::default())
    }
}

pub fn intern_string_intern(s: &String) -> intern_IString {
    STRING_INTERNER.with(|si| intern_IObj{x: si.lock().unwrap().intern(s), phantom: Default::default()})
}

pub fn intern_istring_str(s: &intern_IString) -> String {
    STRING_INTERNER.with(|si| si.lock().unwrap().get(s.x).clone())
}

pub fn intern_istring_ord(s: &intern_IString) -> u32 {
    s.x
}



impl fmt::Display for intern_IString {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        record::format_ddlog_str(&intern_istring_str(self), f)
    }
}


impl fmt::Debug for intern_IString {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        record::format_ddlog_str(&intern_istring_str(self), f)
    }
}

impl serde::Serialize for intern_IString {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where S: serde::Serializer
    {
        intern_istring_str(self).serialize(serializer)
    }
}

impl<'de> serde::Deserialize<'de> for intern_IString {
    fn deserialize<D>(deserializer: D) -> Result<intern_IString, D::Error>
        where D: serde::Deserializer<'de>
    {
        String::deserialize(deserializer).map(|s|intern_string_intern(&s))
    }
}


impl FromRecord for intern_IString {
    fn from_record(val: &Record) -> Result<Self, String> {
        String::from_record(val).map(|s|intern_string_intern(&s))
    }
}

impl IntoRecord for intern_IString {
    fn into_record(self) -> Record {
        intern_istring_str(&self).into_record()
    }
}

impl Mutator<intern_IString> for Record
{
    fn mutate(&self, s: &mut intern_IString) -> Result<(), String> {
        let mut string = intern_istring_str(s);
        self.mutate(&mut string)?;
        *s = intern_string_intern(&string);
        Ok(())
    }
}
