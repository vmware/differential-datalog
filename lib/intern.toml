[dependencies.lazy_static]
version = "1.3"
