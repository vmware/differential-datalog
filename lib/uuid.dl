/* Universally Unique Identified (UUID) library.
 *
 * Functions in this library are bindings for functions in
 * the Rust uuid crate: https://crates.io/crates/uuid
 */

extern type Uuid

typedef Error = string

/* The nil UUID is special form of UUID that is specified to have all 128 bits 
 * set to zero, as defined in IETF RFC 4122 Section 4.1.7. */
extern function nil(): Uuid

/* Creates a UUID from a 128bit value in big-endian order.
 */
extern function from_u128(v: bit<128>): Uuid

/* Creates a UUID from a 128bit value in little-endian order.
 */
extern function from_u128_le(v: bit<128>): Uuid

/* Creates a UUID using the supplied big-endian bytes.
 * This function will return an error if b has any length other than 16.
 */
extern function from_bytes(b: Vec<bit<8>>): Result<Uuid, Error>

/* Parses a Uuid from a string of hexadecimal digits with optional hyphens.
 * Any of the formats generated by this module (simple, hyphenated, urn) are
 * supported by this parsing function.
 */
extern function parse_str(str: string): Result<Uuid, Error>

/* Convert to lower-case hyphenated format:
 * "936da01f-9abd-4d9d-80c7-02af85c822a8".
 */
extern function to_hyphenated_lower(uuid: Uuid): string

/* Convert to upper-case hyphenated format:
 * "936DA01F-9ABD-4D9D-80C7-02AF85C822A8"
 */
extern function to_hyphenated_upper(uuid: Uuid): string

/* Convert to lower-case simple format:
 * "936da01f9abd4d9d80c702af85c822a8".
 */
extern function to_simple_lower(uuid: Uuid): string

/* Convert to upper-case simple format:
 * "936DA01F9ABD4D9D80C702AF85C822A8".
 */
extern function to_simple_upper(uuid: Uuid): string

/* Convert to lower-case urn format:
 * "urn:uuid:936da01f-9abd-4d9d-80c7-02af85c822a8".
 */
extern function to_urn_lower(uuid: Uuid): string

/* Convert to upper-case urn format:
 * "urn:uuid:936DA01F-9ABD-4D9D-80C7-02AF85C822A8".
 */
extern function to_urn_upper(uuid: Uuid): string

/* Returns a 128bit value containing the UUID data.
 */
extern function as_u128(uuid: Uuid): bit<128>

/* Returns a 128bit little-endian value containing the UUID data.
 *
 * The bytes in the UUID will be reversed and packed into a u128. Note that this
 * will produce a different result than to_fields_le, because the entire UUID is
 * reversed, rather than reversing the individual fields in-place.
 */
extern function to_u128_le(uuid: Uuid): bit<128>

/* Tests if the UUID is nil.
 */
extern function is_nil(uuid: Uuid): bool
