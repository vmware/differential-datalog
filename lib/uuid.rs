use differential_datalog::record::Record;
use serde::de::Deserializer;
use serde::ser::Serializer;
use std::default::Default;
use std::fmt;
pub use uuid as crate_uuid;

#[derive(Eq, Ord, Clone, Hash, PartialEq, PartialOrd)]
pub struct Uuid(::uuid::Uuid);

impl Uuid {
    pub fn new(addr: ::uuid::Uuid) -> Self {
        Uuid(addr)
    }
}

impl Deref for Uuid {
    type Target = ::uuid::Uuid;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Default for Uuid {
    fn default() -> Uuid {
        nil()
    }
}

impl fmt::Display for Uuid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", as_u128(self))
    }
}

impl fmt::Debug for Uuid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self, f)
    }
}

impl Serialize for Uuid {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        as_u128(self).serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for Uuid {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        u128::deserialize(deserializer).map(|x| from_u128(&x))
    }
}

impl FromRecord for Uuid {
    fn from_record(val: &Record) -> Result<Self, String> {
        u128::from_record(val).map(|x| from_u128(&x))
    }
}

impl IntoRecord for Uuid {
    fn into_record(self) -> Record {
        as_u128(&self).into_record()
    }
}

impl Mutator<Uuid> for Record {
    fn mutate(&self, uuid: &mut Uuid) -> Result<(), String> {
        Uuid::from_record(self).map(|u| *uuid = u)
    }
}

pub fn nil() -> Uuid {
    Uuid::new(::uuid::Uuid::nil())
}

pub fn nAMESPACE_DNS() -> Uuid {
    Uuid::new(::uuid::Uuid::NAMESPACE_DNS)
}

pub fn nAMESPACE_OID() -> Uuid {
    Uuid::new(::uuid::Uuid::NAMESPACE_OID)
}

pub fn nAMESPACE_URL() -> Uuid {
    Uuid::new(::uuid::Uuid::NAMESPACE_URL)
}

pub fn nAMESPACE_X500() -> Uuid {
    Uuid::new(::uuid::Uuid::NAMESPACE_X500)
}

pub fn new_v5(namespace: &Uuid, name: &ddlog_std::Vec<u8>) -> Uuid {
    Uuid::new(::uuid::Uuid::new_v5(namespace, &**name))
}

pub fn from_u128(v: &u128) -> Uuid {
    Uuid::new(::uuid::Uuid::from_u128(*v))
}

pub fn from_u128_le(v: &u128) -> Uuid {
    Uuid::new(::uuid::Uuid::from_u128_le(*v))
}

pub fn from_bytes(b: &ddlog_std::Vec<u8>) -> ddlog_std::Result<Uuid, Error> {
    ddlog_std::res2std(::uuid::Uuid::from_slice(&**b).map(Uuid::new))
}

pub fn parse_str(s: &String) -> ddlog_std::Result<Uuid, Error> {
    ddlog_std::res2std(::uuid::Uuid::parse_str(&s).map(Uuid::new))
}

pub fn to_hyphenated_lower(uuid: &Uuid) -> String {
    uuid.to_hyphenated_ref()
        .encode_lower(&mut ::uuid::Uuid::encode_buffer())
        .to_string()
}

pub fn to_hyphenated_upper(uuid: &Uuid) -> String {
    uuid.to_hyphenated_ref()
        .encode_upper(&mut ::uuid::Uuid::encode_buffer())
        .to_string()
}

pub fn to_simple_lower(uuid: &Uuid) -> String {
    uuid.to_simple_ref()
        .encode_lower(&mut ::uuid::Uuid::encode_buffer())
        .to_string()
}

pub fn to_simple_upper(uuid: &Uuid) -> String {
    uuid.to_simple_ref()
        .encode_upper(&mut ::uuid::Uuid::encode_buffer())
        .to_string()
}

pub fn to_urn_lower(uuid: &Uuid) -> String {
    uuid.to_urn_ref()
        .encode_lower(&mut ::uuid::Uuid::encode_buffer())
        .to_string()
}

pub fn to_urn_upper(uuid: &Uuid) -> String {
    uuid.to_urn_ref()
        .encode_upper(&mut ::uuid::Uuid::encode_buffer())
        .to_string()
}

pub fn as_u128(uuid: &Uuid) -> u128 {
    uuid.as_u128()
}

pub fn to_u128_le(uuid: &Uuid) -> u128 {
    uuid.to_u128_le()
}

pub fn is_nil(uuid: &Uuid) -> bool {
    uuid.is_nil()
}
