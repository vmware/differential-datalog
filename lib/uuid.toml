[dependencies.uuid]
version = "0.8"
