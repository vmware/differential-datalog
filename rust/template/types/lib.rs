#![allow(
    path_statements,
    unused_imports,
    non_snake_case,
    non_camel_case_types,
    non_upper_case_globals,
    unused_parens,
    non_shorthand_field_patterns,
    dead_code,
    overflowing_literals,
    unreachable_patterns,
    unused_variables,
    clippy::unknown_clippy_lints,
    clippy::missing_safety_doc
)]

use num::bigint::BigInt;
use num::FromPrimitive;
use num_traits::identities::One;
use ordered_float::OrderedFloat;
use std::borrow;
use std::fmt;
use std::hash::Hash;
use std::ops::Deref;
use std::os::raw; // TODO: this is  only used by ovn.rs and should be moved there.
use std::result;

use serde::de::DeserializeOwned;
use serde::Deserialize;
use serde::Serialize;

use lazy_static::lazy_static;

use differential_datalog::ddval::*;
use differential_datalog::decl_enum_into_record;
use differential_datalog::decl_record_mutator_enum;
use differential_datalog::decl_record_mutator_struct;
use differential_datalog::decl_struct_into_record;
use differential_datalog::int::*;
use differential_datalog::program::*;
use differential_datalog::record;
use differential_datalog::record::FromRecord;
use differential_datalog::record::IntoRecord;
use differential_datalog::uint::*;

mod log;
pub use log::*;

/* FlatBuffers code generated by `flatc` */
#[cfg(feature = "flatbuf")]
mod flatbuf_generated;

/* `FromFlatBuffer`, `ToFlatBuffer`, etc, trait declarations. */
#[cfg(feature = "flatbuf")]
pub mod flatbuf;

pub trait Val:
    Eq + Ord + Clone + Hash + PartialEq + PartialOrd + Serialize + DeserializeOwned
{
}

impl<T> Val for T where
    T: Eq + Ord + Clone + Hash + PartialEq + PartialOrd + Serialize + DeserializeOwned
{
}

pub fn string_append_str(mut s1: String, s2: &str) -> String {
    s1.push_str(s2);
    s1
}

#[allow(clippy::ptr_arg)]
pub fn string_append(mut s1: String, s2: &String) -> String {
    s1.push_str(s2.as_str());
    s1
}

#[macro_export]
macro_rules! deserialize_map_from_array {
    ( $modname:ident, $ktype:ty, $vtype:ty, $kfunc:ident ) => {
        mod $modname {
            use super::*;
            use serde::de::{Deserialize, Deserializer};
            use serde::ser::Serializer;
            use std::collections::BTreeMap;

            pub fn serialize<S>(
                map: &__std::std_Map<$ktype, $vtype>,
                serializer: S,
            ) -> Result<S::Ok, S::Error>
            where
                S: Serializer,
            {
                serializer.collect_seq(map.x.values())
            }

            pub fn deserialize<'de, D>(
                deserializer: D,
            ) -> Result<__std::std_Map<$ktype, $vtype>, D::Error>
            where
                D: Deserializer<'de>,
            {
                let v = Vec::<$vtype>::deserialize(deserializer)?;
                Ok(v.into_iter().map(|item| ($kfunc(&item), item)).collect())
            }
        }
    };
}

/*- !!!!!!!!!!!!!!!!!!!! -*/
// Don't edit this line
// Code below this point is needed to test-compile template
// code and is not part of the template.
