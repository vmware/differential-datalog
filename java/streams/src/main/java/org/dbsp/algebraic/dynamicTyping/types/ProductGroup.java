/*
 * Copyright (c) 2021 VMware, Inc.
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.dbsp.algebraic.dynamicTyping.types;

import org.dbsp.algebraic.dynamicTyping.DynamicGroup;
import org.dbsp.algebraic.staticTyping.Monoid;
import org.dbsp.lib.LinqIterator;

import java.util.List;

/**
 * A product group is a product of several groups, operating pointwise.
 */
public class ProductGroup implements DynamicGroup {
    final List<DynamicGroup> components;

    public ProductGroup(List<DynamicGroup> components) {
        this.components = components;
    }

    @SuppressWarnings("unchecked")
    @Override
    public Object negate(Object data) {
        List<Object> list = (List<Object>)data;
        return LinqIterator.create(this.components).zip(list).map(p -> p.first.negate(p.second));
    }

    @SuppressWarnings("unchecked")
    @Override
    public Object add(Object left, Object right) {
        List<Object> lList = (List<Object>)left;
        List<Object> rList = (List<Object>)right;
        return LinqIterator.create(this.components)
                .zip3(lList, rList)
                .map(p -> p.first.add(p.second, p.third))
                .toList();
    }

    @Override
    public Object zero() {
        return LinqIterator.create(this.components).map(Monoid::zero).toList();
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("Group<");
        boolean first = true;
        for (DynamicGroup t: this.components) {
            if (!first)
                builder.append(", ");
            first = false;
            builder.append(t.toString());
        }
        builder.append(">");
        return builder.toString();
    }
}
