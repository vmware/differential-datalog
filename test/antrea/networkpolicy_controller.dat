log_level 100000;
echo Adding namespace and pods;
start;

insert k8spolicy.Namespace(
    .name = "ns1",
    .uid = k8spolicy.UID{"e340296d-84e2-48b2-9b7a-91641db60db5"},
    .labels = [("l1", "l1v1")]),

insert k8spolicy.Namespace(
    .name = "ns2",
    .uid = k8spolicy.UID{"37e6f556-02f3-4775-93a7-ec924ee78810"},
    .labels = []),

insert k8spolicy.Pod(
    .name = "ns1.pod1",
    .namespace = "ns1",
    .uid = k8spolicy.UID{"adb006eb-07fd-4a4b-9b09-d899267aeadc"},
    .labels = [],
    .spec = k8spolicy.PodSpec{.nodeName = "node1"},
    .status = k8spolicy.PodStatus{"192.168.1.1"}
),

insert k8spolicy.Pod(
    .name = "ns1.pod2",
    .namespace = "ns1",
    .uid = k8spolicy.UID{"17024561-7949-4f63-a248-bc29075f04b2"},
    .labels = [],
    .spec = k8spolicy.PodSpec{.nodeName = "node2"},
    .status = k8spolicy.PodStatus{"192.168.1.2"}
),

insert k8spolicy.Pod(
    .name = "ns2.pod1",
    .namespace = "ns2",
    .uid = k8spolicy.UID{"2fc100dd-a981-47e2-8a62-e9b2dff3a5d3"},
    .labels = [],
    .spec = k8spolicy.PodSpec{.nodeName = "node1"},
    .status = k8spolicy.PodStatus{"192.168.1.1"}
),

insert k8spolicy.Pod(
    .name = "ns2.pod2",
    .namespace = "ns2",
    .uid = k8spolicy.UID{"ad847460-cb25-43e7-a44c-d2ff7bdb65ac"},
    .labels = [],
    .spec = k8spolicy.PodSpec{.nodeName = "node2"},
    .status = k8spolicy.PodStatus{"192.168.1.2"}
),

insert k8spolicy.Pod(
    .name = "<top>.pod1",
    .namespace = "",
    .uid = k8spolicy.UID{"a5f3a901-040d-407c-a842-918e250f79de"},
    .labels = [],
    .spec = k8spolicy.PodSpec{.nodeName = "node1"},
    .status = k8spolicy.PodStatus{"192.168.1.1"}
),

insert k8spolicy.Pod(
    .name = "<top>.pod2",
    .namespace = "",
    .uid = k8spolicy.UID{"23d84e05-e0b6-446d-a919-440d1b74ebfc"},
    .labels = [],
    .spec = k8spolicy.PodSpec{.nodeName = "node2"},
    .status = k8spolicy.PodStatus{"192.168.1.2"}
),

commit dump_changes;

echo Adding policies;
start;

insert k8spolicy.NetworkPolicy(
    .name = "ns1.policy1",
    .namespace = "ns1",
    .uid = k8spolicy.UID{"156bc79a-ee16-47d9-82b6-8bc48849af74"},
    .spec = k8spolicy.NetworkPolicySpec{
        .podSelector = k8spolicy.LabelSelector{
            .matchLabels = [("l1", "l1v1")],
            .matchExpressions = []
        },
        .ingress = [k8spolicy.NetworkPolicyIngressRule{
                        .ports = [k8spolicy.NetworkPolicyPort{
                                      .protocol = std.Some{"TCP"},
                                      .port = std.None
                                  }],
                        .from = [k8spolicy.NetworkPolicyPeer{
                                    .podSelector = std.Some{
                                        k8spolicy.LabelSelector{
                                            .matchLabels = [("l1", "l1v1")],
                                            .matchExpressions = []
                                        }
                                    },
                                    .namespaceSelector = std.Some{
                                        k8spolicy.LabelSelector{
                                            .matchLabels = [],
                                            .matchExpressions = []
                                        }
                                    },
                                    .ipBlock = std.None
                                 }]
                    }],
        .egress = [],
        .policyType = []
    }
),


commit dump_changes;
