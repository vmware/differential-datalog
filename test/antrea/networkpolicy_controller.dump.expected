Adding namespace and pods
Ok8spolicy.Namespace:
k8spolicy_Namespace{.name = "ns1", .uid = k8spolicy_UID{.uid = "e340296d-84e2-48b2-9b7a-91641db60db5"}, .labels = []}: +1
k8spolicy_Namespace{.name = "ns2", .uid = k8spolicy_UID{.uid = "37e6f556-02f3-4775-93a7-ec924ee78810"}, .labels = []}: +1
Ok8spolicy.Pod:
k8spolicy_Pod{.name = "<top>.pod1", .namespace = "", .uid = k8spolicy_UID{.uid = "a5f3a901-040d-407c-a842-918e250f79de"}, .labels = [], .spec = k8spolicy_PodSpec{.nodeName = "node1"}, .status = k8spolicy_PodStatus{.podIP = "192.168.1.1"}}: +1
k8spolicy_Pod{.name = "<top>.pod2", .namespace = "", .uid = k8spolicy_UID{.uid = "23d84e05-e0b6-446d-a919-440d1b74ebfc"}, .labels = [], .spec = k8spolicy_PodSpec{.nodeName = "node2"}, .status = k8spolicy_PodStatus{.podIP = "192.168.1.2"}}: +1
k8spolicy_Pod{.name = "ns1.pod1", .namespace = "ns1", .uid = k8spolicy_UID{.uid = "adb006eb-07fd-4a4b-9b09-d899267aeadc"}, .labels = [], .spec = k8spolicy_PodSpec{.nodeName = "node1"}, .status = k8spolicy_PodStatus{.podIP = "192.168.1.1"}}: +1
k8spolicy_Pod{.name = "ns1.pod2", .namespace = "ns1", .uid = k8spolicy_UID{.uid = "17024561-7949-4f63-a248-bc29075f04b2"}, .labels = [], .spec = k8spolicy_PodSpec{.nodeName = "node2"}, .status = k8spolicy_PodStatus{.podIP = "192.168.1.2"}}: +1
k8spolicy_Pod{.name = "ns2.pod1", .namespace = "ns2", .uid = k8spolicy_UID{.uid = "2fc100dd-a981-47e2-8a62-e9b2dff3a5d3"}, .labels = [], .spec = k8spolicy_PodSpec{.nodeName = "node1"}, .status = k8spolicy_PodStatus{.podIP = "192.168.1.1"}}: +1
k8spolicy_Pod{.name = "ns2.pod2", .namespace = "ns2", .uid = k8spolicy_UID{.uid = "ad847460-cb25-43e7-a44c-d2ff7bdb65ac"}, .labels = [], .spec = k8spolicy_PodSpec{.nodeName = "node2"}, .status = k8spolicy_PodStatus{.podIP = "192.168.1.2"}}: +1
Adding policies
LOG (0): createAppliedToGroup ns1.policy1
LOG (0): createAppliedToGroup ns1.policy1
AppliedToGroup:
AppliedToGroup{.uid = k8spolicy_UID{.uid = "dbd5cd23-1488-56b6-925e-7354a2189c7d"}, .name = "dbd5cd23-1488-56b6-925e-7354a2189c7d", .selector = types_GroupSelector{.normalizedName = "namespace=ns1 And podSelector=l1 In [l1v1]", .nsSelector = types_NSSelectorNS{.ns = "ns1"}, .podSelector = std_Some{.x = k8spolicy_LabelSelector{.matchLabels = [("l1", "l1v1")], .matchExpressions = []}}}}: +1
NetworkPolicy:
NetworkPolicy{.uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .name = "ns1.policy1", .namespace = "ns1", .rules = [], .appliedToGroups = ["dbd5cd23-1488-56b6-925e-7354a2189c7d"]}: +1
NetworkPolicyExt:
NetworkPolicyExt{.policy = NetworkPolicy{.uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .name = "ns1.policy1", .namespace = "ns1", .rules = [], .appliedToGroups = ["dbd5cd23-1488-56b6-925e-7354a2189c7d"]}, .addressGroups = []}: +1
Ok8spolicy.NetworkPolicy:
k8spolicy_NetworkPolicy{.name = "ns1.policy1", .namespace = "ns1", .uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .spec = k8spolicy_NetworkPolicySpec{.podSelector = k8spolicy_LabelSelector{.matchLabels = [("l1", "l1v1")], .matchExpressions = []}, .ingress = [], .egress = [], .policyTypes = []}}: +1
Making changes
LOG (0): createAppliedToGroup ns1.policy1
LOG (0): createAppliedToGroup ns1.policy1
LOG (0): createAppliedToGroup ns1.policy1
LOG (0): createAppliedToGroup ns1.policy1
AddressGroup:
AddressGroup{.uid = k8spolicy_UID{.uid = "df49fa6b-0346-5dc9-9bb2-3da4fdbe8dd8"}, .name = "df49fa6b-0346-5dc9-9bb2-3da4fdbe8dd8", .selector = types_GroupSelector{.normalizedName = "namespaceSelector= And podSelector=l1 In [l1v1]", .nsSelector = types_NSSelectorLS{.selector = k8spolicy_LabelSelector{.matchLabels = [], .matchExpressions = []}}, .podSelector = std_Some{.x = k8spolicy_LabelSelector{.matchLabels = [("l1", "l1v1")], .matchExpressions = []}}}}: +1
NetworkPolicy:
NetworkPolicy{.uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .name = "ns1.policy1", .namespace = "ns1", .rules = [], .appliedToGroups = ["dbd5cd23-1488-56b6-925e-7354a2189c7d"]}: -1
NetworkPolicy{.uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .name = "ns1.policy1", .namespace = "ns1", .rules = [types_NetworkPolicyRule{.direction = types_DirectionIn{}, .from = types_NetworkPolicyPeer{.addressGroups = ["df49fa6b-0346-5dc9-9bb2-3da4fdbe8dd8"], .ipBlocks = []}, .to = types_NetworkPolicyPeer{.addressGroups = [], .ipBlocks = []}, .services = [types_Service{.protocol = "TCP", .port = std_None{}}]}], .appliedToGroups = ["dbd5cd23-1488-56b6-925e-7354a2189c7d"]}: +1
NetworkPolicyAddressGroup:
NetworkPolicyAddressGroup{.np = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .addressGroup = "df49fa6b-0346-5dc9-9bb2-3da4fdbe8dd8"}: +1
NetworkPolicyExt:
NetworkPolicyExt{.policy = NetworkPolicy{.uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .name = "ns1.policy1", .namespace = "ns1", .rules = [], .appliedToGroups = ["dbd5cd23-1488-56b6-925e-7354a2189c7d"]}, .addressGroups = []}: -1
NetworkPolicyExt{.policy = NetworkPolicy{.uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .name = "ns1.policy1", .namespace = "ns1", .rules = [types_NetworkPolicyRule{.direction = types_DirectionIn{}, .from = types_NetworkPolicyPeer{.addressGroups = ["df49fa6b-0346-5dc9-9bb2-3da4fdbe8dd8"], .ipBlocks = []}, .to = types_NetworkPolicyPeer{.addressGroups = [], .ipBlocks = []}, .services = [types_Service{.protocol = "TCP", .port = std_None{}}]}], .appliedToGroups = ["dbd5cd23-1488-56b6-925e-7354a2189c7d"]}, .addressGroups = [AddressGroup{.uid = k8spolicy_UID{.uid = "df49fa6b-0346-5dc9-9bb2-3da4fdbe8dd8"}, .name = "df49fa6b-0346-5dc9-9bb2-3da4fdbe8dd8", .selector = types_GroupSelector{.normalizedName = "namespaceSelector= And podSelector=l1 In [l1v1]", .nsSelector = types_NSSelectorLS{.selector = k8spolicy_LabelSelector{.matchLabels = [], .matchExpressions = []}}, .podSelector = std_Some{.x = k8spolicy_LabelSelector{.matchLabels = [("l1", "l1v1")], .matchExpressions = []}}}}]}: +1
Ok8spolicy.Namespace:
k8spolicy_Namespace{.name = "ns1", .uid = k8spolicy_UID{.uid = "e340296d-84e2-48b2-9b7a-91641db60db5"}, .labels = []}: -1
k8spolicy_Namespace{.name = "ns1", .uid = k8spolicy_UID{.uid = "e340296d-84e2-48b2-9b7a-91641db60db5"}, .labels = [("l1", "l1v1")]}: +1
Ok8spolicy.NetworkPolicy:
k8spolicy_NetworkPolicy{.name = "ns1.policy1", .namespace = "ns1", .uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .spec = k8spolicy_NetworkPolicySpec{.podSelector = k8spolicy_LabelSelector{.matchLabels = [("l1", "l1v1")], .matchExpressions = []}, .ingress = [], .egress = [], .policyTypes = []}}: -1
k8spolicy_NetworkPolicy{.name = "ns1.policy1", .namespace = "ns1", .uid = k8spolicy_UID{.uid = "156bc79a-ee16-47d9-82b6-8bc48849af74"}, .spec = k8spolicy_NetworkPolicySpec{.podSelector = k8spolicy_LabelSelector{.matchLabels = [("l1", "l1v1")], .matchExpressions = []}, .ingress = [k8spolicy_NetworkPolicyIngressRule{.ports = [k8spolicy_NetworkPolicyPort{.protocol = std_Some{.x = "TCP"}, .port = std_None{}}], .from = [k8spolicy_NetworkPolicyPeer{.podSelector = std_Some{.x = k8spolicy_LabelSelector{.matchLabels = [("l1", "l1v1")], .matchExpressions = []}}, .namespaceSelector = std_Some{.x = k8spolicy_LabelSelector{.matchLabels = [], .matchExpressions = []}}, .ipBlock = std_None{}}]}], .egress = [], .policyTypes = []}}: +1
